src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/CertificateStore.cs

```diff
@@ -66,16 +66,18 @@ public CertificateStore()
         /// Find all available certificates
         /// </summary>
         /// <returns>Sequence of certificates (possibly empty).</returns>
-        public IEnumerable<X509Certificate2> FindAll()
-        {
-            var query =
-                from name in _storeNames
-                from location in _storeLocations
-                from cert in FindAll(name, location)
-                select cert;
-            var result = query.ToList();
-            return result;
-        }
+        /// <remarks>
+        /// This does not return *distinct* certificates. If the same certificate appears in different stores/locations
+        /// it can be returned multiple times.
+        /// We intentionally avoid filtering out duplicates at this stage because the same certificate might appear
+        /// *with* a private key in one location, and *without* one in another location. Performing a 'Distinct()' here
+        /// might arbitrarily exclude the one with a private key, making it appear as if there is no private key for it.
+        /// </remarks>
+        public IEnumerable<X509Certificate2> FindAll() =>
+            from name in _storeNames
+            from location in _storeLocations
+            from cert in FindAll(name, location)
+            select cert;
```

Comment:
Needs to include `ToList()` to ensure all the enumeration is completed up front.

Reply:
But it returns `IEnumerable` so callers shouldn't have any expectation that it's enumerated up front.

Putting in a `ToList()` just means the collection ends up being enumerated twice. The second time will be much less costly, so no big deal, but if we're doing that why not explicitly return an `IList`?

Reply:
Deferred execution of `IEnumerable` still traps folks - it's a recurring theme on StackOverflow, for example - so I'd prefer to keep things simple for potential consumers. 

I wouldn't change it to return an `IList` because the API should be determined by the consumer's needs, not implementation details.

Reply:
(last stand before changing it...)

But as a consumer (who's not trusted to handle deferred execution) an `IList` would suit my needs better than `IEnumerable` because I wouldn't have to `ToList` it.

I know there's an argument for keeping the interface as generic as possible to give us the freedom to change the implementation, but if we're catering to maintainers of the method itself (not just consumers of it) the type signature serves as a valuable indicator that _they_ shouldn't be returning a deferred enumerable.

This has tripped me up a couple of times, and I think this unwritten rule is at least as much of a problem as the risk of consumers not being able to handle deferred execution.

Reply:
Key problem is that `IList<T>` is read/write - it's inviting consumers to change the results, which we don't want. This is a primary reason why I prefer to expose `IEnumerable<T>`, because it makes it clear that the user doesn't get to modify the list. 

Not a biggie though, so I'll merge this in.